net4j/com.b2international.snowowl.eventbus/src/com/b2international/snowowl/internal/eventbus/EventBus.java

```diff
@@ -74,23 +81,39 @@ protected void doActivate() throws Exception {
 
 	@Override
 	public IEventBus send(String address, Object message) {
-		return send(address, message, null);
+		return send(address, message, IMessage.DEFAULT_TAG);
+	}
+	
+	@Override
+	public IEventBus send(String address, Object message, String tag) {
+		return send(address, message, tag, null);
 	}
 	
 	@Override
 	public IEventBus send(String address, Object message, IHandler<IMessage> handler) {
-		return sendMessageInternal(null, MessageFactory.createMessage(address, message), true, handler);
+		return send(address, message, IMessage.DEFAULT_TAG, handler);
+	}
+
+	@Override
+	public IEventBus send(String address, Object message, String tag, IHandler<IMessage> replyHandler) {
+		return sendMessageInternal(null, MessageFactory.createMessage(address, message, tag), true, replyHandler);
 	}
 	
 	@Override
 	public IEventBus publish(String address, Object message) {
-		return sendMessageInternal(null, MessageFactory.createMessage(address, message), false, null);
+		return publish(address, message, IMessage.DEFAULT_TAG);
+	}
+	
+	@Override
+	public IEventBus publish(String address, Object message, String tag) {
+		return sendMessageInternal(null, MessageFactory.createMessage(address, message, tag), false, null);
 	}
 	
 	@Override
 	public IEventBus receive(IMessage message) {
 		CheckUtil.checkArg(message instanceof BaseMessage, "Accepts only BaseMessage instances");
 		receiveMessage((BaseMessage) message);
+		queue(message);
```

Comment:
This method should be invoked before sending the message for processing in the `doReceive()` method and not here.